src/chapters/style/documentation.typ

```diff
@@ -0,0 +1,247 @@
+Doc comments may not be interrupted by empty lines, markup, or regular comments, i.e. a blokc of documenation must be one continuous sequence of lines starting with three forward slashes `///`.
+
+#do-dont[
+  ```typst
+  /// A1
+  /// A2
+  /// A3
+
+  // discuraged, but valid
+  ///B1
+  ///B2
+  ///B3
+  ```
+][
+  ```typst
+  /// A1
+  //
+  /// B1
+
+  /// A1
+
+  /// B1
+
+  /// C1
+  Hello World
+  /// D1
+
+  /**
+   * This is a regular comment, not a doc comment.
+   */
+  #let item = { ... }
+
+  // This is a regular comment, not a doc comment.
+  #let item = { ... }
+  ```
+]
+
+There are two kinds of documentation comments, inner and outer doc comments.
+Outer doc comments are placed right above the declaration they are attached to.
+
+#do-dont[
+  ```typst
+  /// Documentation for func
+  #let func() = { ... }
+
+  ```
+][
+  ```typst
+  /// Stray doc comment, not the doc for func
+
+  #let func() = {
+    /// Stray doc comment, not the doc for func
+    ...
+  }
+
+  /// Stray doc comment, not the doc for func
+  ```
+]
+
+Inner doc comments are used to document modules and must not have a declaration, instead they refer to the file they are placed in and may only be declared once and as the first non-comment item in a file.
+There may be regular comments before doc comments to allow placing internal documentation or license information at the top of the file.
+Inner doc comments are currently not permitted anywhere else.
+
+#do-dont[
+  ```typst
+  // optional leading comments
+
+  /// Module doc
+
+  /// Function or value doc
+  #let item = { ... }
+  ```
+][
+  ```typst
+  /// Function or value doc
+  #let item = { ... }
+
+  /// Stray doc comment, not the module or func doc
+  ```
+]
+
+Outer doc comments may be used on `let` bindings only.
+
+#do-dont[
+  ```typst
+  /// Function doc
+  #let func() = { ... }
+
+  /// Value doc
+  #let value = { ... }
+  ```
+][
+  ```typst
+  /// Stray doc comment, markup may not be documented
+  Hello World
+
+  /// Stray doc comment, imports my not be documented
+  #import "module.typ"
+
+  /// Stray doc comment, scopes may not be documented
+  #[
+    ...
+  ]
+  ```
+]
+
+Doc comments contain a description of the documented item itself, as well as an optional semantic trailer.
+The content of descriptions should generally be simple Typst markup and should not contain any scripting, (i.e. no loops, conditionals or function calls, except for `#link("...")[...]`), this allows the documentation to be turned into markdown or plain text for language servers to send to editors.
+
+== Description
+As mentioned before, the description should be simple, containing mostly markup and no scripting.
+
+== Semantic Trailer
+The semantic trailer is fully optional, starts with an empty doc comment line to separate it from the description, and may contain:
+- multiple `term` items for parameter type hints and descriptions,
```

Comment:
Do we stick to the term items (so `/` instead of `-`)? I'd also welcome opinions from @Myriad-Dreamin @SillyFreak and @JamesxX on that :)

Reply:
it makes sense imo. My only concern is: is there any chance that we'd need colons as part of the term? For type annotations (which are of course a temporary issue), I would say this could become necessary:
```
/// / entry (key: str, value: str): a single entry, given as a `dict`
```
this would have to become
```
/// / entry: (key: str, value: str) a single entry, given as a `dict`
```
which imo is not as pleasant to read in the code

Reply:
We can have
```
/ entry (key\: str, value\: str): a single entry, given as a `dict`
```

![image](https://github.com/typst-community/guidelines/assets/35292584/858b310f-d6c1-4af3-bdbe-4deede066a15)

But this should be not perfect anymore.

Reply:
Imho only the type belongs there (so `dictionary`) and not the keys it has. The latter should be described in the parameter description. Also, this will probably not be supported with upcoming built-in type annotations (at least I haven't seen any suggestions regarding this) and so it won't work anyway.

Reply:
Perhaps this is best solved by using comments per parameter (as suggested before and on the discord) and then simply using `-> types` for these parameters too?

```typst
/// ...
#let func(
  /// A single entry, given as a `dict` with the keys `key` and `value` of type `str`.
  /// -> dictionary
  entry,
) = { ... }
```

Or with a destructuring pattern (up for debate):

```typst
/// ...
#let func(
  /// A single entry.
  /// -> (key: str, value: str)
  entry,
) = { ... }
```

I'll touch on this in the TLDR below.